File: 1151-minimum-swaps-to-group-all-1s-together/1151-minimum-swaps-to-group-all-1s-together.py
class Solution:
    def minSwaps(self, data: List[int]) -> int:
        ones = 0
        for num in data:
            ones += num
        if ones <= 1:
            return 0
        minSwap = float('inf')
        left = 0
        right = left + ones - 1
        currZeroCount = 0
        for i in range(left, right + 1):
            if data[i] == 0:
                currZeroCount += 1
        minSwap = min(minSwap, currZeroCount)
        while right < len(data):
            oldHead = data[left]
            left += 1
            right += 1
            if right < len(data):
                newTail = data[right]
                if oldHead == 0:
                    currZeroCount -= 1
                if newTail == 0:
                    currZeroCount += 1
                minSwap = min(minSwap, currZeroCount)
        return minSwap

File: 1151-minimum-swaps-to-group-all-1s-together/1151-minimum-swaps-to-group-all-1s-together.java
class Solution {
    public int minSwaps(int[] data) {
        int oneCount = 0;
        int n = data.length;
        int[] memo = new int[n];
        memo[0] = (data[0] == 0) ? 0 : 1;
        for (int i = 1; i < n; i++) {
            memo[i] = memo[i - 1] + ((data[i] == 1) ? 1 : 0);
        }
        oneCount = memo[n - 1];
        int minMoves = Integer.MAX_VALUE;
        int windowSize = oneCount;
        if (windowSize == 0) return 0;
        for (int right = oneCount - 1; right < n; right++) {
            int left = right - oneCount + 1;
            int zeroCount = 0;
            if (left == 0) {
                zeroCount = windowSize - memo[right];
            } else {
                zeroCount = windowSize - memo[right] + memo[left - 1];
            }
            minMoves = Math.min(minMoves, zeroCount);
        }
        return minMoves;
        
    }
}